File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        for(int i=0;i<n;i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            int j = i+1, k = n-1;
            while(j<k){
                if(nums[i]+nums[j]+nums[k]==0){
                    res.add(List.of(nums[i],nums[j],nums[k]));
                    j++;
                    k--;
                    while(j<k && nums[j]==nums[j-1]){
                        j++;
                    }
                    while(j<k && nums[k]==nums[k+1]){
                        k--;
                    }
                } else if(nums[i]+nums[j]+nums[k] < 0){
                    j++;
                    while(j<k && nums[j]==nums[j-1]){
                        j++;
                    }
                }else{
                    k--;
                    while(j<k && nums[k]==nums[k+1]){
                        k--;
                    }
                }
            }
        }
        return res;
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        """
        Same as 2Sum => nums[i]+nums[j] = -1*nums[k] ===> target=(-1*nums[k])
        Loop in for target and then do 2Sum.
        """
        nums.sort()
        n = len(nums)
        result = []
        for i in range(n):
            target = -1* nums[i]
            low, high = i+1, n-1
            if i > 0 and nums[i] == nums[i-1]:
                continue
            while low<high:
                add_n = nums[low]+nums[high]
                if add_n == target:
                    result.append([nums[i], nums[low],nums[high]])
                    while low<high and nums[low]==nums[low+1]:
                        low += 1
                    while low<high and nums[high]==nums[high-1]:
                        high -= 1
                    
                    low += 1
                    high -= 1

                elif add_n>target:
                    high -= 1
                else:
                    low += 1
        return result
